Source/Particles/Deposition/CurrentDeposition.H

```diff
@@ -1533,6 +1533,670 @@ void doChargeConservingDepositionShapeNImplicit (const amrex::ParticleReal * con
+            const auto i_new = static_cast<int>(x_new-shift);
+            int cell_crossings_x = std::abs(i_new-i_old);
+            num_segments += cell_crossings_x;
+
+            // compute cell crossings in Z-direction
+            const auto k_old = static_cast<int>(z_old-shift);
+            const auto k_new = static_cast<int>(z_new-shift);
+            int cell_crossings_z = std::abs(k_new-k_old);
+            num_segments += cell_crossings_z;
+
+            // need to assert that the number of cell crossings in each direction
+            // is within the range permitted by the number of guard cells
+            // e.g., if (num_segments > 3) ...
+
+            // compute total change in particle position and the initial cell
+            // locations in each direction used to find the position at cell crossings.
+            const double dxp = x_new - x_old;
+            const double dzp = z_new - z_old;
+            const auto dirX_sign = static_cast<int>(dxp < 0. ? -1 : 1);
+            const auto dirZ_sign = static_cast<int>(dzp < 0. ? -1 : 1);
+            double Xcell = 0., Zcell = 0.;
+            if (num_segments > 1) {
+                Xcell = i_old + shift + 0.5_rt*(1-dirX_sign);
+                Zcell = k_old + shift + 0.5_rt*(1-dirZ_sign);
+            }
+
+            // loop over the number of segments and deposit
+            Compute_shape_factor< depos_order-1 > compute_shape_factor_cell;
+            Compute_shape_factor_pair< depos_order > compute_shape_factors_node;
+            double dxp_seg, dzp_seg;
+            double x0_new, z0_new;
+            double x0_old = x_old;
+            double z0_old = z_old;
+
+            for (int ns=0; ns<num_segments; ns++) {
+
+                if (ns == num_segments-1) { // final segment
+
+                    x0_new = x_new;
+                    z0_new = z_new;
+                    dxp_seg = x0_new - x0_old;
+                    dzp_seg = z0_new - z0_old;
+
+                }
+                else {
+
+                    x0_new = Xcell + dirX_sign;
+                    z0_new = Zcell + dirZ_sign;
+                    dxp_seg = x0_new - x0_old;
+                    dzp_seg = z0_new - z0_old;
+
+                    if (dzp == 0. || std::abs(dxp_seg) < std::abs(dxp/dzp*dzp_seg)) {
+                        Xcell = x0_new;
+                        dzp_seg = dzp/dxp*dxp_seg;
+                        z0_new = z0_old + dzp_seg;
+                    }
+                    else {
+                        Zcell = z0_new;
+                        dxp_seg = dxp/dzp*dzp_seg;
+                        x0_new = x0_old + dxp_seg;
+                    }
+
+                }
+
+                // compute the segment factors (each equal to dt_seg/dt for nonzero dxp, or dzp)
+                const auto seg_factor_x = static_cast<amrex::Real>(dxp == 0. ? 1. : dxp_seg/dxp);
+                const auto seg_factor_z = static_cast<amrex::Real>(dzp == 0. ? 1. : dzp_seg/dzp);
```

Comment:
We usually use `amrex::ParticleReal` for particle quanitites and `amrex::Real` for field quantities.

For depositions, there is a natural shift from one to another, just need to pick where.

Reply:
Still a bit confused on types, but it looks like from previous deposition schemes that anything to do with computing the shape factor is kept double. I see the seg_factor terms as part of the shape factor, so I will make them all double.

Reply:
Yes, we have not carefully debugged this yet and promoted some types based on tests we did.

I think more stable would be to make sure the cell edges and position changes dx, dy and dz are in kernels scaled to 0:1 or -1:1 and then we scale up again at the current deposition atomic.